sycl/test-e2e/InvokeSimd/Feature/invoke_simd_struct_by_pointer.cpp

```diff
@@ -0,0 +1,169 @@
+ */
+#ifdef IMPL_SUBGROUP
+#define SUBGROUP_ATTR
+#else
+#define SUBGROUP_ATTR [[intel::reqd_sub_group_size(VL)]]
+#endif
+
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+constexpr int VL = 16;
+
+struct multipliers {
+  int x;
+  int y;
+};
+
+/* Performs the addition of vectors A + B scaled by scalars->x and scalars->y. */
+__attribute__((always_inline)) esimd::simd<int, VL>
+ESIMD_CALLEE(int *A, esimd::simd<int, VL> b, int i,
+             multipliers *scalars) SYCL_ESIMD_FUNCTION {
+  esimd::simd<int, VL> a;
+  a.copy_from(A + i);
+  // Add vectors A + B and scale them by x and y.
+  return a + b * scalars->x * scalars->y;
+}
+
+[[intel::device_indirectly_callable]] SYCL_EXTERNAL
+    simd<int, VL> __regcall SIMD_CALLEE(int *A, simd<int, VL> b, int i,
+                                        multipliers *scalars)
+        SYCL_ESIMD_FUNCTION;
+
+int SPMD_CALLEE(int *A, int b, int i, multipliers *scalars) {
+  return A[i] + b * scalars->x * scalars->y;
+}
+
+using namespace sycl;
+
+constexpr bool use_invoke_simd = true;
+
+int main(void) {
+  constexpr unsigned Size = 1024;
+  constexpr unsigned GroupSize = 4 * VL;
+
+  auto q = queue{gpu_selector_v};
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<sycl::info::device::name>()
+            << "\n";
+  auto ctx = q.get_context();
+
+  int *A = static_cast<int *>(malloc_shared(Size * sizeof(int), dev, ctx));
+  int *B = static_cast<int *>(malloc_shared(Size * sizeof(int), dev, ctx));
+  int *C = static_cast<int *>(malloc_shared(Size * sizeof(int), dev, ctx));
+
+  for (unsigned i = 0; i < Size; ++i) {
+    A[i] = B[i] = i;
+    C[i] = -1;
+  }
+
+  // USM shared memory allocation for a struct multipliers.
+  multipliers *scalars = static_cast<multipliers *>(
+      malloc_shared(sizeof(multipliers), dev, ctx));
+  scalars->x = 2;
+  scalars->y = 3;
+
+  sycl::range<1> GlobalRange{Size};
+  // Number of workitems in each workgroup.
+  sycl::range<1> LocalRange{GroupSize};
+
+  sycl::nd_range<1> Range(GlobalRange, LocalRange);
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Test>(Range, [=](nd_item<1> ndi) SUBGROUP_ATTR {
+        sub_group sg = ndi.get_sub_group();
+        group<1> g = ndi.get_group();
+        uint32_t i =
+            sg.get_group_linear_id() * VL + g.get_group_linear_id() * GroupSize;
+        uint32_t wi_id = i + sg.get_local_id();
+        int res;
+
+        if constexpr (use_invoke_simd) {
+          res = invoke_simd(sg, SIMD_CALLEE, uniform{A}, B[wi_id], uniform{i},
+                            uniform{scalars});
+        } else {
+          res = SPMD_CALLEE(A, B[wi_id], wi_id, scalars);
+        }
+        C[wi_id] = res;
+      });
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    sycl::free(scalars, ctx);
+    sycl::free(A, ctx);
+    sycl::free(B, ctx);
+    sycl::free(C, ctx);
```

Comment:
Or maybe even use buffer to avoid this handling alltogether. We can get a pointer from the accessor.